### PR TITLE
[Dy2stat] Fix Memory Optimization in run_program_op and Add SimNet as Unit Test

### PR DESCRIPTION
**Pull request by zhhsplendid**
### PR types

Others
### PR changes

Others
### Describe

Add Similarity Net as unit test. During the unit test, we found three problems:

1. The run_program_op has memory optimization error when running dy2stat net multiple times.
2. The support for SelectedRows can cause problem in dy2stat.
3. The return grammar has problem.

This PR fixes the 1. problem but modify codes for the 2. 3. problems to make PR smaller. I will fix those two problems in the next PR(s)

### CODE REVIEW

**Review by liym27 — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/run_program_op.h [hunk 1]
```diff
@@ -149,14 +153,44 @@ static void ShareVarsFromScope(const std::vector<Variable *> &vars,
...
+    std::vector<std::string> *skip_vars) {
+  const framework::BlockDesc &block = program.Block(0);
+  const std::vector<framework::OpDesc *> &all_ops = block.AllOps();
+
+  std::unordered_set<std::string> grad_op_output;
+  std::unordered_set<std::string> grad_op_input;
+  for (const framework::OpDesc *op : all_ops) {
+    if (!boost::algorithm::ends_with(op->Type(), "_grad")) {
+      continue;
+    }
```
Comment (Aurelius84) [inline paddle/fluid/operators/run_program_op.h:174]: I have  a question:  here we use  `_grad` to determine whether a op is `grad_op`, but some op's grad_op is not with `_grad` suffix, such `arrary_write`'s grad_op is `array_read`.

**Review by chenwhql — DISMISSED** [review]
LGTM

**Review by Aurelius84 — DISMISSED** [review]
LGTM for the rest

File: paddle/fluid/operators/run_program_op.h (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/operators/run_program_op.h:174]: maybe also can use `op_role`?

Comment (zhhsplendid) [reply paddle/fluid/operators/run_program_op.h:174]: Thank you for great suggestion. Done

**Review by chenwhql — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]

**Review by liym27 — APPROVED** [review]

**Pull request merged**